devito/types/sparse.py

```diff
@@ -931,61 +1035,157 @@ class PrecomputedSparseFunction(AbstractSparseFunction):
+        # to setup the coordinates data later on
+        npoint = kwargs.get('npoint', None)
+        if self.npoint and coordinates is None and gridpoints is None:
+            coordinates = np.zeros((npoint, self.grid.dim))
+
+        if coordinates is not None:
+            self._coordinates = self.__subfunc_setup__(coordinates, 'coords')
+            self._gridpoints = None
+            self._dist_origin = {self._coordinates: self.grid.origin_offset}
+        else:
+            assert gridpoints is not None
+            self._coordinates = None
+            self._gridpoints = self.__subfunc_setup__(gridpoints, 'gridpoints')
+            self._dist_origin = {self._gridpoints: self.grid.origin_ioffset}
+
+        # Setup the interpolation coefficients. These are compulsory
+        interpolation_coeffs = kwargs.get('interpolation_coeffs',
+                                          kwargs.get('interpolation_coeffs_data'))
+        self._interpolation_coeffs = \
+            self.__subfunc_setup__(interpolation_coeffs, 'interp_coeffs')
+        self._dist_origin.update({self._interpolation_coeffs: None})
+        # Make sure it matches the radius
+        if self._interpolation_coeffs.shape[-1] != r:
+            nr = self._interpolation_coeffs.shape[-1]
+            raise ValueError("Interpolation coefficients shape %d do "
+                             "not match specified radius %d" % (r, nr))
+
+        warning("Ensure that the provided interpolation coefficient and grid "
+                "point values are computed on the final grid that will be used "
+                "for other computations.")
+
+        self.interpolator = PrecomputedInterpolator(self)
+
+    @property
+    def r(self):
+        return self._radius
+
+    @cached_property
+    def _point_increments(self):
+        """Index increments in each dimension for each point symbol."""
+        return tuple(product(range(-self.r//2+1, self.r//2+1), repeat=self.grid.dim))
+
+    @cached_property
+    def _point_support(self):
+        return np.array(self._point_increments)
+
+    @property
+    def _coords_indices(self):
+        if self.gridpoints is not None:
+            return self.gridpoints.data._local
+        else:
+            return (
+                np.floor(self.coordinates_data - self.grid.origin) / self.grid.spacing
+            ).astype(np.int32)
 
     @property
     def gridpoints(self):
         return self._gridpoints
 
     @property
-    def interpolation_coeffs(self):
-        """ The Precomputed interpolation coefficients."""
-        return self._interpolation_coeffs
+    def gridpoints_data(self):
+        try:
+            return self.gridpoints.data._local
+        except AttributeError:
+            return None
 
-    def _dist_scatter(self, data=None):
-        data = data if data is not None else self.data
-        distributor = self.grid.distributor
+    @property
+    def coordinates(self):
```

Comment:
this should be in the superclass? also the one below?

Reply:
done